lib/basket.rb

```diff
@@ -1,38 +1,39 @@
 class Basket
-  attr_reader :product
-  attr_accessor :quantity
+  attr_reader :products
+  ProductNotFound = Class.new(StandardError)
 
   def initialize(products = [])
     @products = products
   end
 
 # later: def add(product, quantity=1)
   def add(product)
-    @products << product
+    products << product
     #add warehouse quantity -= 1 later on
   end
 
-  def delete(id)
-    product_in_basket = @products.find { |prod| prod.id == id }
+  def delete(product_id)
+    product_in_basket = products.find { |prod| prod.id == product_id }
     if product_in_basket
-      @products.delete(product_in_basket)
+      products.delete(product_in_basket)
     else
-      puts "product not found in your basket!"
+      raise ProductNotFound
```

Comment:
Agreed. Changing to:
`  def delete(product_id)
    product_in_basket = products.find { |prod| prod.id == product_id }
    raise ProductNotFound unless product_in_basket
    products.delete(product_in_basket)
  end`
Btw, is the above allright without any empty lines?